smart-freight-center/sfc-backend
# Feat/add companies endpoint

## Description

- This PR adds an endpoint that returns the companies that a user who can send his data
- Uses the public key of each participant to identify them in the sfc unit

### How to test it

- Create a `provider` realm on keyCloak
- Create a `consumer` realm on keyCloak
- Create a `client` named `amazon` on the `provider` realm
- Create a `client` named `intelcom` on the `consumer` realm
- Copy the public key on the provider realm from `Realm Settings > Keys > RSA256 > Public Certificate`. Paste that value as the public key of amazon on `participants.json`
- Copy the public key on the `consumer` realm from `Realm Settings > Keys > RSA256 > Public Certificate`. Paste that value as the public key of `intelcom` on `participants.json`
- Run the 3 servers 
```bash
yarn dev # simulates the SFC unit (runs on PORT 3000)
yarn dev:provider # (runs on PORT 3001) 
yarn dev:consumer # (runs on PORT 3002) 
````

- Hit `{{provider-host}}/auth/token/` to generate a provider token
- Hit `{{consumer-host}}/auth/token/` to generate a consumer token
- Hit  `{{sfc-unit}}/companies` with authorization of `Bearer <token>`
- It should return the user's connectionÏ

## Approach

The problem with this decentralized authentication/authorization is that a user can create any client name. For example, on the amazon instance, amazon could create a client with the `intelcom` client_id. 

This PR solves this issue by using the user's public key to decode the token anytime they want to access endpoints in the SFC unit. This means that during the setup, each user would share their public_key with the SFC unit.

## Learning

- In this flow, keyCloak uses the private key to encode the tokens on each server then we use the public key from keyCloak to decode the tokens. This gives us the certitude that the correct server generated the token. See [this video](https://www.youtube.com/watch?v=GSIDS_lvRv4&ab_channel=Computerphile) to understand this flow more
